### ISSUE TITLE
Circuit Diagram

### ISSUE THREAD

**Issue by kamranshahidbutt** [HerrRiebmann/KDS2Bluetooth#1]
Dear can you please share the circuit diagram how's we need to connect the circuit.

**Comment by aster94** [comment]
Maybe you are looking for this one?

**Comment by kamranshahidbutt** [comment]
Dear i am looking for Schematics

**Comment by HerrRiebmann** [comment]
Hi, it is quite easy to wire up.
Connect Rx to Rx, Tx to Tx and VCC to 5V Output from Arduino. VCC, Gnd and K-Line to the Bike. A PullUp-resistor is needed between power supply (VCC) and K-Line!

It´s an old one I found, but maybe it helps:
[http://fs5.directupload.net/images/161209/axzt2a7a.jpg](url)
![Scematic](http://fs5.directupload.net/images/161209/axzt2a7a.jpg)
EDIT: It seems that I forgot the PullUp-Resistor (510Ohm) between K-Line and VCC.

Anything else can be found here:
[https://github.com/HerrRiebmann/KDS2Bluetooth/blob/master/Documentation/Readme.md](url)

**Comment by Gui13** [comment]
FWIW I have reproduced this schematic but had to reverse RX and TX for the bluetooth:
- HC06 RX connected to D10 and TX connected to D11

Otherwise the connection wouldn't work.

I still couldn't have the device speak with my bike (Versys 650). I could connect to it though.

@HerrRiebmann the pull up is needed between VCC (5V) or VIN (12V)?

**Comment by Gui13** [comment]
Another guy on an American forum has posted what looks like a good version of the circuit:

![Circuit.png](https://s27.postimg.org/cy9gphe03/Circuit.png)

The capacitor is 10nf and the resistor is 510 Ohms.

**Comment by HerrRiebmann** [comment]
Yes, that is the right one!
I think you don´t need to connect the 5V to the L9637D. But it does not matter.
Rx to Rx and Tx to Tx. Normally it has to be twisted, but the L96...-label considered this already.

**Comment by HerrRiebmann** [comment]
Updated the Readme on the [Dokumentation](https://github.com/HerrRiebmann/KDS2Bluetooth/tree/master/Documentation)-Section:
![L9631DSchematic](https://github.com/HerrRiebmann/KDS2Bluetooth/blob/master/Documentation/Schematic%20L9637D.png)

**Comment by Snecer** [comment]
I am newbie and have a question about your new schematic.

I dont understand the "left" side of the shematic, is my schematic below ok? 

![shema1](https://user-images.githubusercontent.com/36166688/35858972-9dc8a624-0b3e-11e8-9eb0-6ca89623c7c8.png)

**Comment by diaslemuel** [comment]
Hey has anyone tried using a ELM327 instead?? Looking for a easy way

**Comment by HerrRiebmann** [comment]
If it easily would be possible just to connect a ELM327 (_or a cheap china clone_), do you think we all would spend so much time and work into that? ;)

Theoretically this might be possible. But you have to reconfigure the initial process, manually. Then change the timings, baud, SID´s, header, format and more. This is possible, but needs to be done via bluetooth somehow. Then you still do not have matching PIDs, nor the calculations will fit.
Whatever device you might connect then, will possibly reset these settings. If not, it will expect OBD II compatible values and ID´s. These cannot be configured on the ELM327.

Long story short: Might be possible, but needs some research and testing. And you cannot use any stock device to receive the data. You have to develop this yourself.

**Comment by diaslemuel** [comment]
I do understand that your project was a task and i did get all the stuff to do it before i read another forum.

I just asked as some on the zx-10r forum managed to get that done and hence asked

FYI - The Links

https://www.zx-10r.net/forum/f23/clearing-ecu-codes-obdii-scanner-349538.html

https://www.zx-10r.net/forum/f23/obdii-zx10r-bluedriver-320554.html

**Comment by HerrRiebmann** [comment]
This won´t work for any bike without EURO 4 homogenisation (_Which started 2017-01-01_).
I guess, that the 20**16** ZX10R was changed to OBD II protocol already.

As you can see in the video from the second link, the PID´s are pretty standard (0x0C for RPM). All Kawasaki´s (_Bikes, Jetski, ATV, ect._) between 2007-2017 use 0x09.
Also clearing the trouble codes would use SID 0x04 in OBD II, but 0x14 in KDS.

It seems, that the bikes now, really support OBD II, and you just need a plug/cable to make an ELM327 work. That was announced by law, but I never got any further information about this (_And if this is only be done in Europe_) until yet :)
Personally I´vent tried that due to the fact, I do not own a bike from 2017.

I got all the stuff here to create a plug for a few cents, without cutting any cable. Maybe I´ll contact my dealer :D

**Comment by Snecer** [comment]
ZX10R is another story, that works for Gen5+ = 2016 models+

https://ftecu.com/shop/kds-to-obdii-adapter-harnsess/

**Comment by diaslemuel** [comment]
The dealers here are not helpful. and I rely on the internet where it becomes even more confusing

**Comment by Snecer** [comment]
![5a996fdf25bab_vezjekds2bluetooth png 6b0f0ca450dc4fe2ac2ae76aef345c7c](https://user-images.githubusercontent.com/36166688/48059537-b9f76e00-e1b9-11e8-827d-115467edb4ef.png)
![versyskds2btjpg thumb jpg c33844e35c07d0ccfd7dcbefcf64319a](https://user-images.githubusercontent.com/36166688/48059538-ba900480-e1b9-11e8-9db2-4ba4b1c6944e.jpg)

another improved working diagram

**Comment by diaslemuel** [comment]
how do i identify the pins of the L9637. The IC is way to small and as far as i see there is no marking of pin no etc

**Comment by Snecer** [comment]
![hpgejat](https://user-images.githubusercontent.com/36166688/48085772-84737480-e1fb-11e8-8615-1952f071b487.png)

upper left is 1, one side has different edge!

![20161214113604_48062](https://user-images.githubusercontent.com/36166688/48085925-e207c100-e1fb-11e8-8802-e0ea6da8a415.jpg)

**Comment by HerrRiebmann** [comment]
> 
> The dealers here are not helpful. and I rely on the internet where it becomes even more confusing

The dealer should not do anything, I want to test it myself with a borrowed bike :)

Exactly, the most L9637D´s does not have the cut in the top. They have a flat edge on one side.
It is very helpful to buy a SOIC Board, to solder the L9637 on. As you can see underneath @Snecer ´s image.

**Comment by diaslemuel** [comment]
is these the correct resistors etc??
![cart](https://user-images.githubusercontent.com/38744927/48981376-965d7e80-f0fa-11e8-91c2-07b2951c4205.JPG)

**Comment by diaslemuel** [comment]
Do i need to update all 4 .io files to arduino??

**Comment by Snecer** [comment]
![akano](https://user-images.githubusercontent.com/36166688/50549905-c5b95b00-0c65-11e9-9c55-b5d68c0d9294.jpg)

yes, all 4

**Comment by diaslemuel** [comment]
I have uploaded all the 4 sketches one by one (sniffer, kline,elm327 and then ecu_reader)

Is there any other way to upload. Sry but couldnt find any tutorials online

**Comment by Snecer** [comment]
You should open all 4 sketches  (see my screenshot, 4 tabs), connect the device and press button send...

**Comment by diaslemuel** [comment]
Erased the old sketch and uploaded the fresh sketch. The Torque app is connecting to the bike and shows connected to ECU but nothing happens. No rev counter and shows transmission faults. When i try to clear it  the device disconnects and doesnt re-connect unless powered down and on again

**Comment by HerrRiebmann** [comment]
Try to use a Bluetooth Serial Monitor App. Then you can better see what happens!
Maybe you got the wrong Baud-Rate between the BT-Module and Arduino?
Try to use an example sketch, where you can send Data between Serial Monitor an BT-Module.

**Comment by diaslemuel** [comment]
I am using the same code from Github. is it because i didnt use the 20K and 10K ??

**Comment by HerrRiebmann** [comment]
I think so, you have to [pullup](https://en.wikipedia.org/wiki/Pull-up_resistor) the dataline, to straighten the signal.

**Comment by diaslemuel** [comment]
I have added the 20k and 10k as well as per the diagram above but doesnt seem to still read the ECU.

Should i try this instead of the american one shared by snecer??

Fyi - My motorcycle is a ninja 300 non abs
![687474703a2f2f6673352e64697265637475706c6f61642e6e65742f696d616765732f3136313230392f61787a74326137612e6a7067](https://user-images.githubusercontent.com/38744927/53694442-fee08980-3dd4-11e9-86e1-0873ad4e0576.jpg)

**Comment by Snecer** [comment]
![versys 3d](https://user-images.githubusercontent.com/36166688/53694570-186fda00-3db1-11e9-82aa-0f780958cb84.JPG)
![versys pcb](https://user-images.githubusercontent.com/36166688/53694571-19087080-3db1-11e9-9362-fd5a7ec12a98.JPG)

mine diagram is working

![versyskds2btjpg](https://user-images.githubusercontent.com/36166688/53694577-26bdf600-3db1-11e9-96b9-42057a430516.jpg)

**Comment by diaslemuel** [comment]
@Snecer did you get the board made?

**Comment by HerrRiebmann** [comment]
> 
> 
> I have added the 20k and 10k as well as per the diagram above but doesnt seem to still read the ECU.
> 
> Should i try this instead of the american one shared by snecer??
> 
> Fyi - My motorcycle is a ninja 300 non abs
> ![687474703a2f2f6673352e64697265637475706c6f61642e6e65742f696d616765732f3136313230392f61787a74326137612e6a7067](https://user-images.githubusercontent.com/38744927/53694442-fee08980-3dd4-11e9-86e1-0873ad4e0576.jpg)

I think you have to change RX & TX. Instead of other chips, L9637D is already twisted. So connect Rx to Rx and Tx to Tx. I wrote this correctly, but it was wrong in the (old) picture. Sorry about that!

Like @Snecer, I also connected them directly. Nice solution to connect both boards btw. :)

**Comment by HerrRiebmann** [comment]
My current solution looks like that and it is capable of Kawasaki and with a different plug (the grey one), also Suzuki.
![ecu modul sml](https://user-images.githubusercontent.com/15670869/53796962-7decd380-3f35-11e9-928c-af693ce4ac78.jpg)
Unique PCB with standalone ATmega, but stock HC06 Bluetooth.

Doesn´t got any signal drops anymore, or something similar. Only a little overflow, going 16.000rpm on the Gsx-R...

**Comment by diaslemuel** [comment]
Did you @HerrRiebmann and @Snecer get PCB boards made??

**Comment by diaslemuel** [comment]
@HerrRiebmann @Snecer 

Did you guys design a PCB board?? DO you have any design ready

**Comment by HerrRiebmann** [comment]
Yes we both did. But mine was designed with a friend (_due to my lack of knowledge doing that_) and I do not have the source of the PCB design.
I guess Snecer´s image is very easy to adapt and redesign it yourself.

**Comment by diaslemuel** [comment]
@HerrRiebmann I am facing the same issue when I attempted to design the PCB. I dont have a clue how to do it.

@Snecer - can you share the pcb image or design so i get one printed for me. Due credit will be given.

My attempt at trying to get the DIY board is nowhere hence a PCB with all things assembled may help

**Comment by diaslemuel** [comment]
@Snecer Can you help

**Comment by Snecer** [comment]
PCB was friend's design, will try to find the file.
another simple shema...

![Konektor-zadaj](https://user-images.githubusercontent.com/36166688/56079009-b7471780-5dee-11e9-9331-c014ac0d83fe.jpg)
![shemax](https://user-images.githubusercontent.com/36166688/56079010-b7471780-5dee-11e9-99b3-2fb2e5048e28.png)
![vezjex](https://user-images.githubusercontent.com/36166688/56079011-b7471780-5dee-11e9-95f7-e25205f6bf84.png)
![WP_20180302_010](https://user-images.githubusercontent.com/36166688/56079012-b7471780-5dee-11e9-9d05-363a2abc0333.jpg)
![Uploading WP_20180302_011.jpg…]()

**Comment by Snecer** [comment]
https://youtu.be/XFqLPIQ9M2Y

HW works ok, but it seems that SW is not optimized, i also couldn't find a string for speed, must be different as for z750. Thats not a big issue, because gps has that data.

**Comment by diaslemuel** [comment]
@Snecer Please do let me know if you find it. TY in advance

**Comment by diaslemuel** [comment]
Any Update??

**Comment by Snecer** [comment]
No. Find someone who can draw a PCB.

**Comment by Snecer** [comment]
> @Snecer did you get the board made?

I've found someone, who will draw the PCB.

**Comment by diaslemuel** [comment]
@Snecer that would be great. How do i contact

**Comment by IAmOrion** [comment]
I've made this so far using Fritzing - does it look right? I want to make sure I've not missed anything or connected wrong? Once confirmed I will share Gerber Files to use with JLCPCB or what ever PCB company you want

PS: JP1 and JP2 is a solder bridge - in case someone uses a chip with same layout but RX and TX are switched. So you choose which if "RX" is actually RX or TX, and the same to "TX" of the chip (L9637D)

Yellow tracks are on the top layer of the board, orange tracks are on the bottom layer.

If someone can confirm and let me know the wattage of resistors, I can improve the design with SMD 1206 resistors.  Also capacitor spec - atm the design uses 5mm diameter polarised alu caps

I'm hoping to be able to use this with my BMW Z4 with K-Line

<img width="884" alt="Screenshot 2019-08-31 at 01 32 59" src="https://user-images.githubusercontent.com/3501380/64056967-5d719f00-cb8f-11e9-84da-6dd895ee5ce7.png">

I have made my own PCB about 10 times now for different projects so I am used to using Fritzing to design PCBs

**Comment by Snecer** [comment]
excellent, will ask my friend, he is also working on pcb
![Untitled png 361cb7a2a400f548dc9aae12c9c2c446](https://user-images.githubusercontent.com/36166688/64059574-b822ef00-cbbf-11e9-811a-8489328bc406.png)

**Comment by IAmOrion** [comment]
> excellent, will ask my friend, he is also working on pcb

Thanks - that looks much better than mine!
I obviously don't know the voltage rating for the caps or resistors though hence using Through Hole for first version - if someone can let me know the capacitor voltage requirements and the resistor power requirements I can alter my design also :)

I have Eagle too - although for simplicity I use Fritzing.  Will your friend be willing to share his .brd and .sch files?

**Comment by Snecer** [comment]
> > excellent, will ask my friend, he is also working on pcb
> 
> Thanks - that looks much better than mine!
> I obviously don't know the voltage rating for the caps or resistors though hence using Through Hole for first version - if someone can let me know the capacitor voltage requirements and the resistor power requirements I can alter my design also :)
> 
> I have Eagle too - although for simplicity I use Fritzing. Will your friend be willing to share his .brd and .sch files?

Here some .brd files (alpha version).
https://www100.zippyshare.com/v/V8Iv6KWb/file.html

**Comment by IAmOrion** [comment]
> Here some .brd files (alpha version).
> https://www100.zippyshare.com/v/V8Iv6KWb/file.html

Awesome, thanks.  I was hoping they would contain the values of the resistors etc but it doesn't ha.  Sods law! It shows package details, but the value field has been left blank on all of them.

**Comment by diaslemuel** [comment]
Hey @IAmOrion @Snecer @HerrRiebmann  Has there been any update. Cause I have shelved this DIY cause i dont know where i am going wrong. Probably I have got some diodes/resistors or wiring wrong i dont know. Unfortunately i dont have any friend who can help. Hence was looking forward to the board

**Comment by IAmOrion** [comment]
> Hey @IAmOrion @Snecer @HerrRiebmann Has there been any update. Cause I have shelved this DIY cause i dont know where i am going wrong. Probably I have got some diodes/resistors or wiring wrong i dont know. Unfortunately i dont have any friend who can help. Hence was looking forward to the board

I changed my design as I managed to get a load of TH3122.4 chips off of AliExpress (Although I've yet to test them but others have reported they work as expected)
For my particular usage, I don't need bluetooth BUT if you did, I highly recommend using the Keywish Nano BLE - which is an Arduino Nano with built in Bluetooth

I'm too much of a PCB Design novice for Eagle or EasyEDA - but you're welcome to my Gerbers to get them produced.

<img width="872" alt="Screenshot 2019-09-15 at 12 45 00" src="https://user-images.githubusercontent.com/3501380/64921103-c7588e00-d7b6-11e9-91cc-0490f421efff.png">

JLCPCB will make 5 boards for $2 - shipping will obv depend on your location.  I paid just over £9 GBP to have 30 boards made and posted to the UK economy recorded postage

<img width="930" alt="Screenshot 2019-09-15 at 12 47 30" src="https://user-images.githubusercontent.com/3501380/64921138-41891280-d7b7-11e9-86b7-2dddc14d9ca3.png">

<img width="448" alt="Screenshot 2019-09-15 at 12 48 44" src="https://user-images.githubusercontent.com/3501380/64921141-464dc680-d7b7-11e9-8018-1fd84186611c.png">

<img width="462" alt="Screenshot 2019-09-15 at 12 48 51" src="https://user-images.githubusercontent.com/3501380/64921150-4b127a80-d7b7-11e9-9528-073e2c7473e8.png">

Still waiting for them to arrive :)

Gerbers: 
[BMW K Line v4 (Resler Variant TH3122).zip](https://github.com/HerrRiebmann/KDS2Bluetooth/files/3613720/BMW.K.Line.v4.Resler.Variant.TH3122.zip)

For what it's worth, I *MAY* revisit this using the L9637D but I couldn't find enough info of the circuit required - I mean, I found dozens of circuits, but they never showed the capacitor values and I don't know enough to work out what I would need and I couldn't get anyone to reply to me so I stopped working on it.

This as far as I got with the L9637D (I can't remember is this is complete or not)

<img width="1133" alt="Screenshot 2019-09-15 at 12 55 56" src="https://user-images.githubusercontent.com/3501380/64921220-5fa34280-d7b8-11e9-9c02-9cafabf46ffb.png">

Gerbers:
[BMW K Line.zip](https://github.com/HerrRiebmann/KDS2Bluetooth/files/3613724/BMW.K.Line.zip)

Fritzing:
[BMW K Line.fzz.zip](https://github.com/HerrRiebmann/KDS2Bluetooth/files/3613727/BMW.K.Line.fzz.zip)

**Comment by diaslemuel** [comment]
Ive got the L9637D chip. Sourcing these tiny bits in India is a task altogether. None of the sites have all parts.

**Comment by Mattbatt84** [comment]
So I have built this and connect tot he Bluetooth via Torque light. Could somebody let me know what connector and color wire I connect to on a 2011 Versys 650? I have power ground and have have traced back from my ECU for both the self diag terminal and the Kawasaki Self diag terminal. Is it the 6 pin connector or 4 pin? Cheers

**Comment by Snecer** [comment]
check the link

https://www.alter.si/tema/majster-za-arduino.2635004/page-2



V V sob., 3. jul. 2021 ob 20:39 je oseba Mattbatt84 <
***@***.***> napisala:

**Comment by Mattbatt84** [comment]
Is anyone using the Torque lite Android app? I have successfully connected to Bluetooth but seem to get any further than that. Does the app connect instantly or do you need to wait?

**Comment by Snecer** [comment]
> Is anyone using the Torque lite Android app? I have successfully connected to Bluetooth but seem to get any further than that. Does the app connect instantly or do you need to wait?

I am using Torque Pro.
https://youtu.be/O3VYY08cjY0

**Comment by HerrRiebmann** [comment]
Yes, I also use Torque lite, Car Scan and RaceChrono.
And my Garmin Virb XE, off course. Works fine, with the latest code from here.

Just did a completely rewrote of my code. Now with Suzuki & Kawasaki combined, much more performance and error-logging.
Plenty of settings to use the data raw, from storage, gear calculation, obd2 conform, and so on.
I want to release a lot of it, but I think I can´t publish everything.

So I will create a new repository with at least all the features from this one, but more relyable.
Stay tuned for new videos and a new repository in the next time. I´ll let you know here :)

**Comment by Roughd1433** [comment]
@HerrRiebmann Great news. I currently have two bikes in the works and plan on building my own dashboards for each one. So a Suzuki port of the software is wonderful. Thank you for all your hard work. I might be able to bend metal to my will....code not so much :)

**Comment by Mattbatt84** [comment]
After a lot of thinking my wiring was the issue I have used the emulator to check and all wiring and components are good the issue is the KLine coming from the bike. I even traced the kline back the to ecu based on the manual. Torque lite jut does not pick up that the bike is connected. I have a 2011 versys 650. Has anyone else had similar problems and found a solution?

**Comment by HerrRiebmann** [comment]
@Roughd1433 you can see how it works on my Suzuki. Its much better then on Kawasaki, since I introduced a data storage.
It holds everything which comes at once and transferes the values amazingly fast one by one.
Kawasaki just gives you a single value, so you are limited to ~6 or 7 values per second.
https://www.youtube.com/watch?v=RrPropqd4N4

@Mattbatt84 oh, that reminds me of releasing the latest version of the emulator, as well!
The emulator works directly by serial. So there is no L9637D involved, which might is broken or in the wrong direction?!
I created another Arduino with the L9637D, so I can test with the emulator the same way. But you will need another powersource for the second arduino.
Checking the K-Line otherwise than with this device, is tough! You´d need an oscilloscope and something which triggers the K-Line. Or drive to a workshop, where they just clear your diagnostic codes, or something. This would proove the K-Line is right!
The Versys 650 should be similar to the ER650 or nowadays the Ninja 650. I can try the test that out. But I already know, the plug is slightly different.

**Comment by Mattbatt84** [comment]
@HerrRiebmann You were right it was the L9637D. I must have popped it prototyping a new board. I am going to purchase a few more a well as a MC33199 and have a crack at using that as it has a  transmission Baud rate up to 200kBaud. See if makes a difference.  Will also publish the EASYEDA files once done so people can order from PCBway or JLPCB as there pick and place services will do all the little 0806 bits as the  I will be adding TX RX leds to the kline chip to diagnose nonsense like I have been having.

**Comment by Mattbatt84** [comment]
This project is getting expensive. £30 for the boards £30 Parts £30 in broken prototypes. All to see error codes hahahahha.
![kline](https://user-images.githubusercontent.com/56921839/125501094-7a588623-5dcc-43b2-b204-80675ee80ca1.png)

**Comment by HerrRiebmann** [comment]
Wow! That is really expensive!
I think I paid about 160€ for 20 assembled boards. Just without the HC-06 module and without wiring and plugs.
Did you tried https://jlcpcb.com ?

But don´t ask me about all my prototypes, parts and unassembled boards... I guess this could leave the three-digits ;)
That´s the definition of "hobby"!

But all in all, that board-design really looks good! Quite compact!

**Comment by Mattbatt84** [comment]
I looked at jlpcp but they were quoting £135 for 5 boards pre assembled and 25% of the parts were not available. I went with boards only and parts I do not have from ebay. The costly bits were the ATMEGA32U4-MU £8 and L9637D £5. But I am also being impatient and paying premium for parts that are next day shipping soooooooo. If I was willing to wait I think I would be looking at £40 for 5 assembled boards. 
Compact for a reason. I am going to pot them into encloses. Take full advantage of the ATMEGA and most parts 0603 footprint. Will also see if I can heat sink the whole board as the linear voltage regulator has not been heat efficient at all

**Comment by padilia** [comment]
@Snecer link has expired, could you please share again?

**Comment by nmzr26** [comment]
**Important note about “HC-06”**
**Make sure the HC-06 board you are buying has a CSR BC04 chip on it in the picture otherwise it wont work with this project**
**TL;DR:** Many cheap “HC-06” modules are actually **BLE (BN1211/HM-10)**, not Classic Bluetooth.
If it **shows up in nRF Connect** and **does NOT create a COM port**, it’s **not a real HC-06**.
A real HC-06 uses **Classic BT (SPP)**, creates a **COM port**, and **won’t appear in BLE scanners**.

Many cheap modules sold as HC-06 are actually BLE-only modules (e.g. BN1211 / HM-10 class), not Classic Bluetooth.

How to tell the difference:
Real HC-06 (Classic BT / SPP)
1}Creates a COM port on Windows. Asks for a password when connecting
2)Works with PC serial tools and classic Android apps
3)Does NOT appear in nRF Connect (BLE scanner)

Fake / Mislabelled “HC-06” (BN1211 / BLE)

Appears only in BLE scanners (nRF Connect shows GATT services)
No COM port on PC
Will not work with software expecting SPP / serial
Will not connect to your obd app or show up on pc via bluetooth
This caused me a lot of confusion because BLE modules are often sold as HC-06.
Use NRF connect app If your device only shows up under Bluetooth LE, it is not a real HC-06.

**Comment by HerrRiebmann** [comment]
Thank you for your findings @nmzr26 !
It was already a pain in the a🤬 to deal with all the different modules. Sometimes a linefeed, mandatory timings, then different AT commands, differing initial baud rates, and so on.
To just take HC10 modules and label them to HC06 is really another level!
Good that I still have several working ones around. 

It´s possible to add a serial profile to the HC10. But this is a lot of work.
I also thought it would be a great idea to change to a new ESP32 with BLE and it took ages to just communicate with the Serial Bluetooth Terminal. This is a total different approach with LE 🤷🏻‍♂️

**Comment by nmzr26** [comment]
I just took out the ble hc06 and got a real hc06 from a friend. Though i have been trying 
To get the software working with the original bikediagnostic.ino file, I updated the Baud rate to 9600 and changed the Serial pins to 11 and 10. the connection is unstable, only communicating successfully about 2 out of 25 attempts(real ecu). it worked completly fine when it connected.

Communication Logs (Torque Lite / Car Scanner)
The emulator starts and the ECU initializes correctly, but then fails during the diagnostic request:

ECU Initialization: 81 11 F1 81 04

ECU Initialized: 80 F1 11 03 C1 EA 8F BF

Register Start Diagnosis: 80 11 F1 02 10 80 14 (found)

Start Diagnostics Found: 80 F1 11 02 50 80 54

Keep Alive Request: 80 11 F1 01 3E C1 (Found)

Keep Alive : 80 F1 11 01 7E 01 (Keep Alive 0)

Unknown Register: 80 11 F1 02 01 00 85

Error 0: 80 F1 11 03 7E 21 10 35

The sequence ends here and continues to repeat this error cycle.
As I’m not a programmer by any stretch, I’ve been using ChatGPT to help me tweak the K-Line Fast Init timings, but I still haven't been able to establish a proper connection with the ECU consistantly.(closest i got was when the emmulator said start diagnosis the app says checking model) I know the signal is getting through, however, because I can actually hear the communication pulses vibrating through the sub-throttle actuator, which is acting like a speaker for the data transmission. it got close but gpt decided its going to nuke the whole code so i have to start again now. i just saw you responded so i desided to ask you haha.
what could i be doing wrong??

**Comment by nmzr26** [comment]
> Thank you for your findings [@nmzr26](https://github.com/nmzr26) ! It was already a pain in the a🤬 to deal with all the different modules. Sometimes a linefeed, mandatory timings, then different AT commands, differing initial baud rates, and so on. To just take HC10 modules and label them to HC06 is really another level! Good that I still have several working ones around.
> 
> It´s possible to add a serial profile to the HC10. But this is a lot of work. I also thought it would be a great idea to change to a new ESP32 with BLE and it took ages to just communicate with the Serial Bluetooth Terminal. This is a total different approach with LE 🤷🏻‍♂️

Dont even bother with the hm10 not even worth it. more hassel then ble is worth for this kind of project. for what we are tying to do here( as many apps and laptops dont support ble) hc05 and 06 work fine for it but i dont know how long we can source the csr once before generic takes over.

**Comment by HerrRiebmann** [comment]
That seems like a very good start!
Initialization works, Start Diagnostic and Keepalive. That´s a proper connection to the ECU.
The unknown register is truely an unknown message.
So I guess you are having a Kawasaki? This needs requests like I collected them within this Excel Sheet:
https://github.com/HerrRiebmann/KDS2Bluetooth/blob/master/Documentation/OBD%20II%20PIDs.xlsx
On "KDS PID" you find all the commands, how they look like and mostly with a response.

Having pulses on the actuator is not a good sign! This should not be like that!!!
Please double or triple-check the wiring and that you don´t mess up the ECU or an actuator. 

I more often read about using ChatGPT on my code. It fails all the time. And I don´t really understand why to do this.
Everything is there without changing a single line. You can use the Serial Terminal to setup everything and use the most common functions.
You can sniff a lot of commands and later on check it on the PC.
Or you can use more or less any OBD2 compatible App to do the communication, because the code translates it between Suzuki/Kawasaki and OBD2 for you. That´s why it is quite large and complicated.

My suggestion is: Use the Serial Terminal.
Here are all the Commands to use it:
https://github.com/HerrRiebmann/KDS2Bluetooth/tree/master/Bike_Diagnostic_System#xt-commands---custom-setupfunctions
Understand how it´s going and leave the code untouched at the first place. Later on you can adapt the parts you need. Don´t go "All In" on the first shot. This will never work :)

**Comment by nmzr26** [comment]
Thanks for reply.Yes i have a kawasaki Ex250 fi ninja . I changed the capacitor on the k line to a 10nf cap. I was able to get the ecu reader.ino working correctly by changing MYADDR to 0xF2 and it is connecting and working fine with the emulator and real ecu. But the diagnostic.ino is still giving me the same problems. I tried changing a lot of timings and addresses. Ill try some more things today.
But thanks for making this project i would be so stuck if you didnt make this.
And do you still have the kds pc software?

**Comment by HerrRiebmann** [comment]
That sounds strange. From when is the bike?
Do you already have the "red" six-pin plug for diagnostic?

The timing and the adresses are the same for every K-Line communication. Independent from the model or manufacturer. Even Honda, KTM or Suzuki are accepting 0xF1 as the Tester adress. And stick to the timing.
Only the receiving adress differs between the manufacturers and the control unit. But since the keepalive works, this does not seem to be the issue.

I think you already have an OBD2 compliant bike, which only supports two messages over KWP2000: Keepalive and ECU ID.
Anything else is already covered by the CAN-Bus and you could just attach any ELM327 device and use apps for cars. This would be the case if you have this plug:
https://kds-online.de/en/obd-adapter-euro5-for-motorcycle-6-pin-to-obd2-k-line-can-bus.html

Yes, I do have the KDS software, but it won´t work without the official hardware which I don´t own. But since I reverse engineered more or less anything, there is no real need to have it working.

**Comment by nmzr26** [comment]
It’s the same 4-pin diagnostic connector you worked on. I have a 2008–2012 Ninja 250 FI. I found the 0xF2 value from a post by a user on your Arduino forum (you should really link it here; the explanations there are extremely helpful). From what I’ve been able to research, none of the commercially available OBD scanners (atlest the cheap ones) work with this bike. That’s exactly why I started working on this project.

And the vibrations on the subthrottel actuator was because my battery was going dead.

Everything works with the ecu reader.ino file the only probelm is the bikediagnostic.ino.
Any suggestions would be appriciated.

Update

I just ran the serial sniffer real ecu connected.
XTS will only say ecu true whin i run either XTX or XTZ 
XTIM0 says bypass init
XTIM1 does nothing
ATFI says bus init but doesn't say ok same wih ATSI
[LOG.txt](https://github.com/user-attachments/files/24247537/LOG.txt)
[LOG2.txt](https://github.com/user-attachments/files/24247538/LOG2.txt)


Is there something i am missing here?

**Comment by HerrRiebmann** [comment]
Sorry to come back to you so late. I´ve not seen that you updated the comment with the files.
Those are quite helpful. Because it proves my idea that your bike already is OBD2 compliant and should be accessible via CAN bus.

I got the exact same findings on a Z H2, which only provides rudimental functionality via K-Line such as ECU ID, Keepalive and the DTC. Everything else is answered with a 0x7F which says error.
So in short: The hardware is running great. The bike respondes correctly but does not support KDS anymore. 
You should try to connect an ELM327 to the CAN-Bus and try to use Torque or Car Scan to make a connection.

**Comment by nmzr26** [comment]
Hey, thank you for the response. On the ninja 250fi I have been using it to log data like tps, iap, temp etc using a tweaked (changed how things are calculated) version of ecu reader ino file. 

I haven't been able to figure out the issue with bike diagnostic ino. I will try to connect a elm327 to it. But everything else is working fine using ecu reader. Just not the diagnostic part. 

With the readings i loged I was able to pin point the fault on my bike which was a fuel compliance issue e20 which is not supported by my bike and i cant get the normal e10 or e0. So i bypassed the clutch switch and every is running fine now. Will update if i find anything else.
